woocommerce/woocommerce-android
# Issue/2630 adding grouped products

Fixes part 1/2 of #2630 by supporting the option to add products for a grouped product.

#### Changes
- A new button in the `GroupedProductListFragment` to add products.
- Option to display a list of products (that does not include the current product).
- Added the `recyclerview-selection` library to multi select products and add to the grouped product.
- Added unit tests for `GroupedProductListViewModel` and `ProductSelectionListViewModel`.

#### Screenshots
<img src="https://user-images.githubusercontent.com/22608780/91128469-ec8f9500-e6c5-11ea-80bf-4311503554e2.gif" width="300"/>. <img src="https://user-images.githubusercontent.com/22608780/91128473-eef1ef00-e6c5-11ea-9193-2075ca2d1a79.gif" width="300"/>

#### Notes
- The option to search for a product will be taken care of in another PR.
- Some events are missing for this feature, which will be taken care of in a subsequent PR.
- This PR is in draft till this [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1675) can be reviewed and merged.
- I chose to create new `ProductSelectionListViewModel` for the product selection instead of reusing the existing `ProductListViewModel` since we need to exclude the current product when fetching the product list and return the result to the previous fragment, which is not part of the scope of the `ProductListViewModel`. Rather than reusing the `GroupedProductListViewModel`, I chose to create a new `ProductSelectionListViewModel` since we would need to potentially reuse the same functionality when working on Linked products (upsells and cross sells) in M4 release.

#### Known issues
- There is a bug where if you select an already added grouped product from the list, the grouped products count in Product Detail is incorrect. This was found when testing this subsequent PR and is fixed in that PR.

#### Testing
- Click on a grouped product that has atleast 1 product linked to it. 
  - Notice the grouped product section displayed.
  - Click on the grouped product section and notice the new `Add product` button.
  - Click on the button and notice the list of products displayed. Verify that the grouped product from Step 1 is not visible here.
  - Long press on a product and notice that the product is highlighted and verify that the `Done` menu button is displayed.
  - Click on the "Done" menu button and notice that the selected product is displayed correctly in the grouped product list.
  - Click on the `Done` menu button and notice that the Product detail screen is updated with the correct grouped product count.
  - Click on the "Update" menu button and verify that the product is updated successfully.
- Click on a grouped product without any products linked to it.
  - Notice the grouped product section displayed with the title "Add products to the group".
  - Click on the grouped product section and notice the new `Add product` button.
  - Click on the button and notice the list of products displayed. Verify that the grouped product from Step 1 is not visible here.
  - Long press on a product and notice that the product is highlighted and verify that the `Done` menu button is displayed.
  - Click on the "Done" menu button and notice that the selected product is displayed correctly in the grouped product list.
  - Click on the `Done` menu button and notice that the Product detail screen is updated with the correct grouped product count.
  - Click on the "Update" menu button and verify that the product is updated successfully.
- Click on a grouped product that has atleast 1 product linked to it. 
  - Click on the grouped product section.
  - Click on the `Delete` icon and delete all the products associated with the grouped product.
  - Click on the "Done" menu button and notice that the grouped products section in Product detail, now displays: "Add products to the group".
  - Click on the "Update" menu button and verify that the product is updated successfully.

Update release notes:
